### PR TITLE
ggml: Support OpenMP for multi-thread processing

### PR DESCRIPTION
**Pull request by msy-kato**
Currently, ggml multi-thread processing is  implemented with `pthread_create` and `pthread_join`. This is simple and straightforward, but has the following issues

- Difficult to analyze performance per TID (e.g., `perf`) because each thread is created each time a token is generating.
- Thread affinity and core binding settings are not flexible.

Therefore, I suggest an option using OpenMP, which is general multi-threaded library. it is optional, and by default the previous implementation works.

Below are the results measured on an AWS m7i.24xlarge (Intel Xeon 8488).

```bash
### build
$ cmake -DLLAMA_OPENMP=ON -B build -S .
$ cmake --build build -j$(($(nproc)/2))

### run
$ export OMP_PROC_BIND=TRUE # if need
$ main --model models/llama-2-7b-chat.Q4_0.gguf --temp 0.1 --threads XX --prompt 'AI is going to' --n-predict 256 --seed 0
```

|   Threads   | Pthread<br />[token/sec] | OpenMP<br />[token/sec] | OpenMP<br />OMP_PROC_BIND=TRUE<br />[token/sec] |
| ---- | ------------------------ | ------------------------------------------------ | ----------------------------------------------- |
| 6    | 11.07                    | 11.24                                            | 7.11                                            |
| 12   | 19.28                    | 19.62                                            | 13.37                                           |
| 24   | 26.81                    | 27.11                                            | 22.92                                           |
| 36   | 26.73                    | 28.94                                            | 28.76                                           |
| 42   | 26.08                    | 27.21                                            | 30.54                                           |
| 48   | 25.18                    | 26.72                                            | 30.97                                           |

image

OpenMP version has slightly better performance than the pthread implementation. And setting OMP_PROC_BIND=TRUE improves performance on many threads (Threads >~ 36).

It would be good for users to have additional OpenMP options to tuning for their system.
Any comments are welcome!

related : #7342 #7526

### CODE REVIEW

**Comment by slaren** [comment]
Here is the same test with NKVO that I have been running in the other thread pools PRs:

| GPU         | Model         |   NKVO | Test   |   t/s master |   t/s feat-openmp-support-draft |   Speedup |
|:------------|:--------------|-------:|:-------|-------------:|--------------------------------:|----------:|
| RTX 3090 Ti | llama 7B Q4_0 |      0 | pp512  |      4539.81 |                         4534.23 |      1.00 |
| RTX 3090 Ti | llama 7B Q4_0 |      0 | tg128  |       153.90 |                          155.25 |      1.01 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | pp512  |       502.87 |                          527.40 |      1.05 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | tg128  |        20.85 |                           62.17 |      2.98 |

I have also tested with CPU only, on native Windows with MSVC, and on a Apple MBP with an M3 Max, and I haven't found any performance regressions. So I have to say congratulations, this is the first thread pool attempt that achieves a very large reduction in the thread launch overhead, without any significant downsides. I think we could use OMP by default.

Some concerns:
- The value of `num_threads` is not always respected by OMP, and when this happens the result is a deadlock. The [OMP logic](https://www.openmp.org/spec-html/5.0/openmpsu35.html#x55-880002.6.1) is fairly complicated and it may depend on factors outside of our control. We may need to update the number of threads in the shared state according to the value returned by `omp_get_num_threads`.
- It may be better to avoid using OMP at all if `n_threads == 1`, and instead use the main thread. The OMP implementations that I tested seem to be smart enough to do this already, but I don't think this is guaranteed, and I it would be more reliable to handle this case ourselves and avoid the OMP overhead entirely.

**Comment by netrunnereve** [comment]
> I have also tested with CPU only, on native Windows with MSVC, and on a Apple MBP with an M3 Max, and I haven't found any performance regressions. So I have to say congratulations, this is the first thread pool attempt that achieves a very large reduction in the thread launch overhead, without any significant downsides. I think we could use OMP by default.

On an old 4 core/8 thread Xeon v2 there's a small regression with four threads but it's the same as master with eight. Typically with these threadpool implementations the speedup only happens when you use a lot of threads and thus have to deal with the overhead.
`master`
| model                          |       size |     params | backend    |    threads |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ------------: | ---------------: |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          4 |         pp512 |      4.38 ± 0.00 |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          4 |         tg128 |      3.51 ± 0.00 |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          8 |         pp512 |      4.64 ± 0.02 |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          8 |         tg128 |      3.52 ± 0.00 |

`PR`
| model                          |       size |     params | backend    |    threads |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ------------: | ---------------: |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          4 |         pp512 |      4.34 ± 0.00 |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          4 |         tg128 |      3.30 ± 0.08 |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          8 |         pp512 |      4.64 ± 0.00 |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          8 |         tg128 |      3.52 ± 0.00 |

`PR with OMP_PROC_BIND=TRUE`
| model                          |       size |     params | backend    |    threads |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ------------: | ---------------: |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          8 |         pp512 |      4.64 ± 0.01 |
| llama 7B Q6_K                  |   5.53 GiB |     7.24 B | CPU        |          8 |         tg128 |      3.49 ± 0.00 |

**Comment by slaren** [comment]
I tried with a very small stories260k model, since that should exacerbate the overhead of starting the threads, but it is still consistently faster with this PR (except 1 threads). Is it really a given that waking up a thread needs to be more expensive than starting a thread?

| CPU           | Model                      |   Threads | Test   |   t/s master |   t/s omp |   Speedup |
|:--------------|:---------------------------|----------:|:-------|-------------:|--------------------------------:|----------:|
|  i9-13900K | stories260k F32 |         1 | pp512  |      6920.29 |                         6882.03 |      0.99 |
|  i9-13900K | stories260k F32 |         1 | tg128  |      6937.02 |                         6865.42 |      0.99 |
|  i9-13900K | stories260k F32 |         2 | pp512  |     12365.50 |                        12665.08 |      1.02 |
|  i9-13900K | stories260k F32 |         2 | tg128  |      4687.61 |                         5891.87 |      1.26 |
|  i9-13900K | stories260k F32 |         3 | pp512  |     17501.55 |                        17597.04 |      1.01 |
|  i9-13900K | stories260k F32 |         3 | tg128  |      3988.88 |                         5272.38 |      1.32 |
|  i9-13900K | stories260k F32 |         4 | pp512  |     22587.09 |                        22778.87 |      1.01 |
|  i9-13900K | stories260k F32 |         4 | tg128  |      3454.17 |                         4633.26 |      1.34 |
|  i9-13900K | stories260k F32 |         5 | pp512  |     27142.85 |                        27477.98 |      1.01 |
|  i9-13900K | stories260k F32 |         5 | tg128  |      3028.37 |                         4052.77 |      1.34 |
|  i9-13900K | stories260k F32 |         6 | pp512  |     30735.90 |                        31353.86 |      1.02 |
|  i9-13900K | stories260k F32 |         6 | tg128  |      2669.83 |                         3668.35 |      1.37 |
|  i9-13900K | stories260k F32 |         7 | pp512  |     31723.15 |                        35381.52 |      1.12 |
|  i9-13900K | stories260k F32 |         7 | tg128  |      2388.36 |                         3295.09 |      1.38 |
|  i9-13900K | stories260k F32 |         8 | pp512  |     32668.71 |                        31819.11 |      0.97 |
|  i9-13900K | stories260k F32 |         8 | tg128  |      2140.98 |                         2944.40 |      1.38 |

**Review by slaren — APPROVED** [review]

**Comment by msy-kato** [comment]
@slaren Thank you for the comments and performance test! I have fixed the code to your feedback.

> The value of num_threads is not always respected by OMP, and when this happens the result is a deadlock. The [OMP logic](https://www.openmp.org/spec-html/5.0/openmpsu35.html#x55-880002.6.1) is fairly complicated and it may depend on factors outside of our control. We may need to update the number of threads in the shared state according to the value returned by omp_get_num_threads. Instead, how about limiting the n_threads by omp_get_max_threads() in ggml_graph_plan as in this implementation?

As you say, deadlocks can occur since the number of threads specified in the `num_threads` clause is not always used(It happens especially OMP_DYNAMIC=TRUE). On the other hand, I'm not sure what happens when n_threads in `state_shared` is updated in the parallel section(`omp_get_num_threads()` returns the right value only within the parallel section). 

Instead, how about limiting the number of threads by omp_get_max_threads() in ggml_graph_plan as like [this](https://github.com/ggerganov/llama.cpp/pull/7606/commits/7918ed7f2ca522fb0426e2eb5bbdba28172f8a0d)?  
I have confirmed that deadlock does not occur under the following conditions
- n_threads is greater than the number of physical cores
- n_threads is greater than thread-limit()

Without this modification, deadlock occurred in the above cases.

> It may be better to avoid using OMP at all if n_threads == 1, and instead use the main thread. The OMP implementations that I tested seem to be smart enough to do this already, but I don't think this is guaranteed, and it would be more reliable to handle this case ourselves and avoid the OMP overhead entirely. This case happens when fully offloading a model to a GPU backend, so it is fairly common.

Thanks, I checked [here](https://www.openmp.org/spec-html/5.0/openmpse14.html#x54-800002.6).
> The thread that encountered the parallel construct becomes the master thread of the new team, with a thread number of zero for the duration of the new parallel region. 

If n_threads is 1, main thread is the master of OMP, so the cost of waking up the thread  may not be a problem. However, the modification is simple and it is possible to branch only when n_threads=1.

**Comment by slaren** [comment]
I changed it to update `n_threads` from `omp_get_num_threads` in a `omp single` section, because I don't think that we can rely on the value returned by `omp_get_max_threads`. I also removed the call to `omp_get_thread_limit` since it was causing issues with MSVC, and I think it is not strictly necessary.

**Comment by msy-kato** [comment]
Thanks for the correction!

**Comment by slaren** [comment]
@ggerganov PTAL and let's merge this if you don't have any objections, it will significantly improve performance in some cases and it will allow me to continue work on #6210.

File: ggml.c [hunk 1]
```diff
@@ -19661,6 +19666,59 @@ struct ggml_cplan ggml_graph_plan(const struct ggml_cgraph * cgraph, int n_threa
     return cplan;
 }
 
+static enum ggml_status ggml_graph_compute_parallel(struct ggml_compute_state * workers, int n_threads){
```
Comment (ggerganov) [inline ggml.c:19669]:
```suggestion
static enum ggml_status ggml_graph_compute_parallel(struct ggml_compute_state * workers, int n_threads) {
```

**Review by ggerganov — APPROVED** [review]
Sorry for the delay - I have missed the request for review

**Comment by xatier** [comment]
We need to install `libgomp1` in `server.Dockerfile`, otherwise it won't start.

`apt install libgomp1`

https://github.com/ggerganov/llama.cpp/blob/master/.devops/server.Dockerfile#L19

**Comment by xatier** [comment]
Fixed in https://github.com/ggerganov/llama.cpp/pull/7780

**Comment by BodhiHu** [comment]
Fyi, this changes throws "omp.h not found error" on Ubuntu 22.04.

**Pull request merged**